platform/macos/display_server_macos.mm

```diff
@@ -3056,9 +3056,25 @@
 }
 
 void DisplayServerMacOS::release_rendering_thread() {
+#if defined(GLES3_ENABLED)
+	if (gl_manager_angle) {
+		gl_manager_angle->release_current();
+	}
+	if (gl_manager_legacy) {
+		gl_manager_legacy->release_current();
+	}
+#endif
 }
 
 void DisplayServerMacOS::make_rendering_thread() {
+#if defined(GLES3_ENABLED)
+	if (gl_manager_angle) {
+		gl_manager_angle->make_current();
+	}
+	if (gl_manager_legacy) {
+		gl_manager_legacy->make_current();
+	}
+#endif
```

Comment:
This seems to be called from the thread when it is started and set the current GL context (to the context of the active window), but GL context is already set from `_draw_viewport()` at the start of each window drawing, so it's likely redundant (probably part of the old 3.x single windows code).

Reply:
That seems to be the case indeed. However, I'm not clear on what the right API would be because we now have the concepts of windows and threads somewhat conflated. I'm wondering if the right call would be the following:
- Remove `DisplayServer::make`/`release_rendering_thread()` .
- Remove `make`/release`_current()` from the various GL context classes.
- Modify `window_make_current()` so that taking `INVALID_WINDOW_ID` means releasing the context.

(When I say _remove_, it may be needed to be understood as _deprecate until we can remove_.)

Reply:
In the next push, I'm replacing the commit where I added those by one that simplifies the APIs (without breaking compat, I think). `DisplayServer` as well as the various GL managers can therefore now have a window-specific make current, a global release, and call it a day.

Reply:
For the reference:

Multithreaded compatibility renderer was crashing (in the `GLES3::Utilities::get_video_adapter_api_version()`) before this PR. With it, it's no longer crashing, but still displaying only black screen. Multithreaded Vulkan renderer also not rendering anything (both with and without this PR), might be macOS specific issue, I'll check it later.

Reply:
In both cases, it deadlocks inside `blit_render_targets_to_screen`.

```
Main thread:

   🔻 Waiting for CommandQueueMT.
std::__1::recursive_mutex::lock() (@std::__1::recursive_mutex::lock():7)
CommandQueueMT::Command1<RendererCanvasCull, void (RendererCanvasCull::*)(RID), RID>* CommandQueueMT::allocate_and_lock<CommandQueueMT::Command1<RendererCanvasCull, void (RendererCanvasCull::*)(RID), RID>>() (core/templates/command_queue_mt.h:362)
void CommandQueueMT::push<RendererCanvasCull, void (RendererCanvasCull::*)(RID), RID>(RendererCanvasCull*, void (RendererCanvasCull::*)(RID), RID) (core/templates/command_queue_mt.h:407)
RenderingServerDefault::canvas_item_clear(RID) (servers/rendering/rendering_server_default.h:916)
CanvasItem::_redraw_callback() (scene/main/canvas_item.cpp:137)
CallQueue::_call_function(Callable const&, Variant const*, int, bool) (core/object/message_queue.cpp:221)
CallQueue::flush() (core/object/message_queue.cpp:326)
SceneTree::process(double) (scene/main/scene_tree.cpp:537)
    🔻 Inside main evnet loop.
Main::iteration() (main/main.cpp:4013)
OS_MacOS::run() (platform/macos/os_macos.mm:778)
main (platform/macos/godot_main_macos.mm:84)
start (@start:593)
```


```
Rendering thread:

    🔻 Since nothing seems directy holding it in the Engine code, I assue it's waiting for the next main event loop iteration, but main thread is locked.
_dlock_wait (@_dlock_wait:17)
_dispatch_thread_event_wait_slow (@_dispatch_thread_event_wait_slow:17)
-[CAMetalLayer(MoltenVK) screenMVK] (@-[CAMetalLayer(MoltenVK) screenMVK]:31)
MVKSwapchain::initSurfaceImages(VkSwapchainCreateInfoKHR const*, unsigned int) (@MVKSwapchain::initSurfaceImages(VkSwapchainCreateInfoKHR const*, unsigned int):153)
MVKSwapchain::MVKSwapchain(MVKDevice*, VkSwapchainCreateInfoKHR const*) (@MVKSwapchain::MVKSwapchain(MVKDevice*, VkSwapchainCreateInfoKHR const*):183)
MVKDevice::createSwapchain(VkSwapchainCreateInfoKHR const*, VkAllocationCallbacks const*) (@MVKDevice::createSwapchain(VkSwapchainCreateInfoKHR const*, VkAllocationCallbacks const*):15)
vkCreateSwapchainKHR (@vkCreateSwapchainKHR:31)
RenderingDeviceDriverVulkan::swap_chain_resize(RenderingDeviceDriver::CommandQueueID, RenderingDeviceDriver::SwapChainID, unsigned int) (drivers/vulkan/rendering_device_driver_vulkan.cpp:2663)
RenderingDevice::screen_prepare_for_drawing(int) (servers/rendering/rendering_device.cpp:3190)
RendererCompositorRD::blit_render_targets_to_screen(int, BlitToScreen const*, int) (servers/rendering/renderer_rd/renderer_compositor_rd.cpp:37)
RendererViewport::draw_viewports(bool) (servers/rendering/renderer_viewport.cpp:861)
RenderingServerDefault::_draw(bool, double) (servers/rendering/rendering_server_default.cpp:89)
    🔻 This lock CommandQueueMT mutex.
CommandQueueMT::_flush() (core/templates/command_queue_mt.h:377)
CommandQueueMT::flush_all() (core/templates/command_queue_mt.h:423)
RenderingServerDefault::_thread_loop() (servers/rendering/rendering_server_default.cpp:347)
...
```

Reply:
Probably not a solution, but MT renderer works with the following changes, until something in the main thread will call `RenderingServer` in a wrong time:

```diff
diff --git a/core/templates/command_queue_mt.h b/core/templates/command_queue_mt.h
index 71030d86d1..76b25d75e4 100644
--- a/core/templates/command_queue_mt.h
+++ b/core/templates/command_queue_mt.h
@@ -366,12 +366,17 @@ class CommandQueueMT {
 
 	void _flush() {
 		lock();
+		uint64_t flush_read_ptr_tmp = flush_read_ptr;
+		LocalVector<uint8_t> command_mem_tmp = command_mem;
+		command_mem.clear();
+		flush_read_ptr = 0;
+		unlock();
 
 		WorkerThreadPool::thread_enter_command_queue_mt_flush(this);
-		while (flush_read_ptr < command_mem.size()) {
-			uint64_t size = *(uint64_t *)&command_mem[flush_read_ptr];
-			flush_read_ptr += 8;
-			CommandBase *cmd = reinterpret_cast<CommandBase *>(&command_mem[flush_read_ptr]);
+		while (flush_read_ptr_tmp < command_mem_tmp.size()) {
+			uint64_t size = *(uint64_t *)&command_mem_tmp[flush_read_ptr_tmp];
+			flush_read_ptr_tmp += 8;
+			CommandBase *cmd = reinterpret_cast<CommandBase *>(&command_mem_tmp[flush_read_ptr_tmp]);
 
 			SyncSemaphore *sync_sem = cmd->get_sync_semaphore();
 			cmd->call();
@@ -379,20 +384,16 @@ class CommandQueueMT {
 				sync_sem->sem.post(); // Release in case it needs sync/ret.
 			}
 
-			if (unlikely(flush_read_ptr == 0)) {
+			if (unlikely(flush_read_ptr_tmp == 0)) {
 				// A reentrant call flushed.
-				DEV_ASSERT(command_mem.is_empty());
+				DEV_ASSERT(command_mem_tmp.is_empty());
 				unlock();
 				return;
 			}
 
-			flush_read_ptr += size;
+			flush_read_ptr_tmp += size;
 		}
 		WorkerThreadPool::thread_exit_command_queue_mt_flush();
-
-		command_mem.clear();
-		flush_read_ptr = 0;
-		unlock();
 	}
 
 	void wait_for_flush();
diff --git a/platform/macos/display_server_macos.mm b/platform/macos/display_server_macos.mm
index 854ee59ed8..42cf3b1ce8 100644
--- a/platform/macos/display_server_macos.mm
+++ b/platform/macos/display_server_macos.mm
@@ -2158,7 +2158,7 @@ void DisplayServerMacOS::window_set_size(const Size2i p_size, WindowID p_window)
 }
 
 Size2i DisplayServerMacOS::window_get_size(WindowID p_window) const {
-	_THREAD_SAFE_METHOD_
+	//_THREAD_SAFE_METHOD_
 
 	ERR_FAIL_COND_V(!windows.has(p_window), Size2i());
 	const WindowData &wd = windows[p_window];
diff --git a/platform/macos/godot_content_view.mm b/platform/macos/godot_content_view.mm
index 93bba84783..9d76fded26 100644
--- a/platform/macos/godot_content_view.mm
+++ b/platform/macos/godot_content_view.mm
@@ -55,9 +55,11 @@
 - (void)displayLayer:(CALayer *)layer {
 	DisplayServerMacOS *ds = (DisplayServerMacOS *)DisplayServer::get_singleton();
 	if (OS::get_singleton()->get_main_loop() && ds->get_is_resizing() && need_redraw) {
-		Main::force_redraw();
-		if (!Main::is_iterating()) { // Avoid cyclic loop.
-			Main::iteration();
+		if ((OS::get_singleton()->get_render_thread_mode() != OS::RENDER_SEPARATE_THREAD)) {
+			Main::force_redraw();
+			if (!Main::is_iterating()) { // Avoid cyclic loop.
+				Main::iteration();
+			}
 		}
 		need_redraw = false;
 	}
```

There seems to be a lot of various conditions multithreaded renderer might lock up, it's also constantly crashing in with out of bounds access in the texture storage. Since there's a "WARNING: The Multi-Threaded rendering thread model is experimental. However, bear in mind that at the moment it can lead to project crashes or instability." warning, I'll assume it is expected to be completely broken right now.

Reply:
Checked it with 4.2.1, Vulkan MT renderer seems to be working, OpenGL is crashing, I'll try bisecting to see what broke it later.

Reply:
I only really tested on Windows. I had to change how locking was working in DisplayServer to prevent deadlocks. I tried to propagate the changes across the other platforms, but maybe I didn't it well. I may test on Mac myself to see what's going on.

Reply:
Bisected it to https://github.com/godotengine/godot/pull/87340 - Finish splitting functionality of the `RenderingDevice` backends into `RenderingDeviceDriver` (it's big, so probably won't be easy to detect what exactly is causing the issue).

Reply:
@bruvzg I realized a missing unlock in my changes to MacOS. Would you please re-test? If it deadlocks, please test with #90470 as well.

Reply:
Neither is working.

Since it was originally broken by https://github.com/godotengine/godot/pull/87340, which is not changing `CommandQueueMT` in any way, it's probably not directly related, and the issue is in `RenderingDevice`.

In some cases (resizing window for example), it still lockups with the aforementioned patch (and lockups only involve `RenderingDevice` not `CommandQueueMT`).

Reply:
Alright. Thanks for your help! I'll delve.

Reply:
The more I look into this, the more issues with multi-threading rendering (physics seem OK) I'm seeing. My suggestion (CC @akien-mga) is that we merge this as it is, given it seems it doesn't break single-threading, and I'll make a subsequent PR fixing MT rendering.